File: android/app/src/main/java/ai/thetahealth/mirobody/data/net/ErrorBus.kt
package ai.thetahealth.mirobody.data.net

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow

/**
 * App-wide error stream. ViewModels emit; a single collector at the root (typically
 * MainActivity) renders to a Snackbar. Extra buffer + DROP_OLDEST guarantees emit
 * never suspends — slow consumers just lose the oldest pending error.
 */
class ErrorBus {
    private val _errors = MutableSharedFlow<AppError>(
        extraBufferCapacity = 8,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )
    val errors: SharedFlow<AppError> = _errors.asSharedFlow()

    fun emit(error: AppError) {
        _errors.tryEmit(error)
    }

    fun emit(throwable: Throwable) {
        emit(throwable.toAppError())
    }
}

File: ios/Mirobody/Data/Net/ErrorBus.swift
import Combine
import Foundation

/// App-wide error stream — mirrors `data/net/ErrorBus.kt`. View models emit; a
/// single collector at the root (`RootView`) renders a transient toast. The
/// subject drops events with no subscriber, so a slow/absent consumer never
/// blocks the emitter.
final class ErrorBus: ObservableObject {
    let errors = PassthroughSubject<AppError, Never>()

    func emit(_ error: AppError) {
        errors.send(error)
    }

    func emit(_ throwable: Error) {
        errors.send(throwable.toAppError())
    }
}
